c/misra/src/rules/RULE-21-26/TimedlockOnInappropriateMutexType.ql

```diff
@@ -0,0 +1,74 @@
+/**
+ * @id c/misra/timedlock-on-inappropriate-mutex-type
+ * @name RULE-21-26: The Standard Library function mtx_timedlock() shall only be invoked on mutexes of type mtx_timed
+ * @description The Standard Library function mtx_timedlock() shall only be invoked on mutex objects
+ *              of appropriate mutex type.
+ * @kind path-problem
+ * @precision high
+ * @problem.severity error
+ * @tags external/misra/id/rule-21-26
+ *       correctness
+ *       concurrency
+ *       external/misra/c/2012/amendment4
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+import semmle.code.cpp.dataflow.new.DataFlow
+
+class MutexTimed extends EnumConstant {
+  MutexTimed() { hasName("mtx_timed") }
+}
+
+class MutexInitCall extends FunctionCall {
+  Expr mutexExpr;
+  Expr mutexTypeExpr;
+
+  MutexInitCall() {
+    getTarget().hasName("mtx_init") and
```

Comment:
Could `mtx_init` be a macro?

Reply:
It isn't in glibc or in clang stdc, but it is on FreeBSD.

Should we assume that if it is a macro, it expands to a simple wrapper function, e.g., `#define mtx_init(a, b) __f((a), (b))`? FWIW, that [isn't the case](https://github.com/bitrig/bitrig/blob/c22c5c6ebcc96092f69737815c84927b83e701be/sys/arch/arm64/include/mutex.h#L53) on FreeBSD, which passes `mutexTypeExpr` into a function call for the second arg.

Reply:
Ok, let's keep it like this for now - we'll determine whether we need to expand coverage during compiler compatibility testing.